File: src/ddf/action.py
"""Actions on form."""
from collections import OrderedDict

from django import forms

from .js import JsDictMixin


class Action(JsDictMixin):
    """An action to take on a list of fields."""
    js_attrs = ['conditions']

    def __init__(self, *conditions):
        """Instanciate with a list of condition to require."""
        self.conditions = conditions

    def execute(self, field):
        """Prevent the action from being applied if conditions don't pass."""
        for condition in self.conditions:
            if not condition.validate(field.form):
                return

        self.apply(field)
        return self

    def apply(self, field):
        """Actual application of the action on the form field."""
        raise NotImplemented()

    def unapply(self, form):
        """Revert apply() action."""


class Remove(Action):
    """Remove fields from a form."""

    def apply(self, field):
        """Remove the field and data from field.form."""
        self.original_keys = list(field.form.fields.keys())
        self.original_field = field.form.fields.pop(field.name)
        attr = self.original_field.widget.attrs.get('class', '')
        attr += ' ddf-hide'
        self.original_field.widget.attrs['class'] = attr

    def unapply(self, form):
        """Restore the field and data in field.form."""
        original_field = getattr(self, 'original_field', None)
        if not original_field:
            return
        fields = OrderedDict()
        for key in self.original_keys:
            if key == original_field.name:
                fields[key] = original_field
            elif key in form.fields:
                fields[key] = form.fields[key]
        form.fields = fields


class RemoveChoices(Action):
    """Remove choices from a choice field."""
    js_attrs = ['conditions', 'choices']

    def __init__(self, choices, *conditions):
        """List of choice values to remove if conditions pass."""
        self.choices = choices
        super(RemoveChoices, self).__init__(*conditions)

    def apply(self, field):
        """Raise a ValidationError if the field value is in self.values."""
        value = field.form.data.get(field.name, None)
        if value in self.choices:
            raise forms.ValidationError(
                field.error_messages['invalid_choice'],
                code='invalid_choice',
                params={'value': value},
            )

File: src/ddf/action.js
import debug from 'debug'

var log = debug('ddf.action')

class Action {
  constructor(conditions) {
    this.conditions = conditions || []
  }

  execute(field) {
    let method = 'apply'

    for (var i=0; i<this.conditions.length; i++) {
      if (!this.conditions[i].validate(field.form)) {
        method = 'unapply'
        break
      }
    }

    log(this, '.', method, '(', field, ')')
    this[method](field)
  }
}

// Remove a field from a form.
class Remove extends Action {
  // Hide the field.
  apply(field) {
    field.hide()
  }

  // Show the field.
  unapply(field) {
    field.show()
  }
}

// Remove given choices from a field.
class RemoveChoices extends Action {
  constructor(conditions, choices) {
    super(conditions)
    this.choices = choices
  }

  // Hide options which are not in this.choices from a field.
  apply(field) {
    if (this.choices.indexOf(field.value) >= 0) {
      field.valueClear()
    }

    let options = field.element.querySelectorAll('option')
    for (let i=0; i < options.length; i++) {
      if (this.choices.indexOf(options[i].value) >= 0) {
        options[i].classList.add('ddf-hide')
      }
    }
  }

  // Show options which are not in this.choices from a field.
  unapply(field) {
    for (let i=0; i < options.length; i++) {
      if (!this.choices.indexOf(options[i].value)) {
        options[i].classList.remove('ddf-hide')
      }
    }
  }
}

export {
  Action,
  Remove,
  RemoveChoices
}
